### ISSUE TITLE
ParseFromString() expects bytes and not bytesarray object

### ISSUE THREAD

**Issue by joelkny97** [waymo-research/waymo-open-dataset#885]
The Frame object from dataset_pb2.open_dataset in the tutorial.ipynb is unable to process the bytearray of the tutorial dataset frames downloaded for the tutorial purposes and throws an error saying that the ParseFromString method expects a bytes and not bytearray object. The error message is posted below -

ERROR -
6     frame = open_dataset.Frame()
----> 7     frame.ParseFromString(bytearray(data.numpy()))
8     break
9 

TypeError: expected bytes, bytearray found

**Comment by kaanguler14** [comment]
Filename path may be wrong.

**Comment by ZhiyuanZorian** [comment]
have you solved this

**Comment by Master-cai** [comment]
@ZhiyuanZorian @kaanguler14 @joelkny97 Hiya, have u figured it out?

**Comment by li199603** [comment]
@joelkny97 @Master-cai @kaanguler14 @ZhiyuanZorian Hiya, have u figured it out?

**Comment by Master-cai** [comment]
> [@joelkny97](https://github.com/joelkny97) [@Master-cai](https://github.com/Master-cai) [@kaanguler14](https://github.com/kaanguler14) [@ZhiyuanZorian](https://github.com/ZhiyuanZorian) Hiya, have u figured it out?

```git
- frame.ParseFromString(bytearray(data.numpy()))
+ frame.ParseFromString(bytes(data.numpy()))
```
I can't remember clearly, maybe this change solved my problem.

**Comment by SheepDoctor** [comment]
> > [@joelkny97](https://github.com/joelkny97) [@Master-cai](https://github.com/Master-cai) [@kaanguler14](https://github.com/kaanguler14) [@ZhiyuanZorian](https://github.com/ZhiyuanZorian) Hiya, have u figured it out?
> 
> ```
> - frame.ParseFromString(bytearray(data.numpy()))
> + frame.ParseFromString(bytes(data.numpy()))
> ```
> 
> I can't remember clearly, maybe this change solved my problem.

The python package I download seems to cause this problem too. What I did was to replace all 'bytearray' with 'bytes' in frame_utils.py and it worked out. Whenever the issue happened, replace the 'bytearray' with 'bytes' is always a simple yet efficient approach to solve the problem.  

By the way, I am using waymo-open-dataset-tf-2-12-0==1.6.4 and tensorflow==2.12.0

**Comment by johannes-do** [comment]
Yeah, that works. But still all the waymo functions also have the issue. e.g. parse_range_image_and_camera_projection():

expected bytes, bytearray found

There are some solutions to this:
- Downgrade protobuf to 3.20
- If you can not cause some other package needs a newer protobuf (3.20 is ancient!). You can do:
  - Rebuild a new python package with updated proto (didnt try, not sure if that works out of the box)
  - "Patch" all the bytearrays with bellow code:

```python
from typing import Any

from google.protobuf.message import Message
from google.protobuf.pyext._message import RepeatedScalarContainer


def fix_bytearrays(obj: Any) -> Any:
    """
    Recursively walk a protobuf message (or container) and convert any
    bytearray fields into bytes, in-place. Returns the same object.
    """
    if isinstance(obj, Message):
        # Iterate through all fields
        for field, value in obj.ListFields():
            if isinstance(value, bytearray):
                setattr(obj, field.name, bytes(value))
            elif isinstance(value, list) or isinstance(value, RepeatedScalarContainer):
                for i in range(len(value)):
                    if isinstance(value[i], bytearray):
                        value[i] = bytes(value[i])
                    elif isinstance(value[i], Message):
                        fix_bytearrays(value[i])
            elif isinstance(value, Message):
                fix_bytearrays(value)

    elif isinstance(obj, list):
        for i in range(len(obj)):
            if isinstance(obj[i], bytearray):
                obj[i] = bytes(obj[i])
            elif isinstance(obj[i], Message):
                fix_bytearrays(obj[i])

    return obj
```
